net: lib: http_server: implement concurrent and backlog

MeisterBob opened the pull request:
Currently, the configured values for concurrent and backlog are only used for display in the shell. This PR implements the use of these values.

mrodgers-witekio reviewed (changes requested):
The `backlog` side of things looks good to me.

For the `concurrent` parameter, is there a scenario where we would want this to be lower than the `CONFIG_HTTP_SERVER_MAX_CLIENTS`? I guess maybe if you want to limit the number of client instances used by one service, to keep some free for other services? I'm just trying to understand the use-case better.

If this is the case then I think some changes are needed to the HTTP server sample (and perhaps any other samples using the HTTP server API). The sample currently has a value of 1 for the concurrent parameter. When visiting the page from a web browser using HTTP/1, it opens multiple connections to the server - the effect of this when I gave it a quick try seems to be that the websocket connection won't work (because a second connection won't be accepted) and the network stats aren't shown. So at the very least we should increase the `concurrent` parameter for the sample, probably best to set it equal to `CONFIG_HTTP_SERVER_MAX_CLIENTS`?

Also worth considering that this change could lead to some unexpected breakages of any downstream code that's been based on the sample app and also set the (previously unused) `concurrent` parameter to 1 - so would probably need a migration guide / release note comment too.

rlubos reviewed (commented):
Generally no objections, but I agree we should revisit `HTTP_SERVICE_DEFINE()` and bump the `_concurrent` parameter value as suggested. A migration guide entry would be highly desired as well.

MeisterBob commented:
I updated the samples to accept `CONFIG_HTTP_SERVER_MAX_CLIENTS` connections and added `BUILD_ASSERT`s to the `HTTP[S]_SERVICE_DEFINE[_EMPTY]` macros. Also added a line to the migration guide.

mrodgers-witekio reviewed (dismissed):
Looks good to me now - assuming that the CI is OK (looks like it may just need re-running?)

rlubos reviewed (commented):
Looks that https://github.com/zephyrproject-rtos/zephyr/blob/86a52d925435a89d676bbd0941156cb926fa8f26/tests/net/lib/http_server/common/src/main.c#L53 needs to be adjusted as well, lgtm otherwise.

jukkar commented:
Could you split this to separate commits, so something like this
* the core HTTP changes
* sample changes
* doc change
* test changes

kartben commented:
@MeisterBob please rebase and resolve the merge conflict. Thanks!

mt-gjoerges committed:
samples: net: http_server: concurrent connections

update to reflect changes in 3ebecd15615197b81efe79fcde6e4bedc0f6b09d

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

mt-gjoerges committed:
net: lib: http_server: implement concurrent

only accept new connections until the configured value for concurrent
connections is reached. Also set the backlog of the listening socket
to the configured value.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

mt-gjoerges committed:
tests: net: http_server: concurrent connections

update to reflect changes in 3ebecd15615197b81efe79fcde6e4bedc0f6b09d

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

mt-gjoerges committed:
doc: migration-guide-4-2: http_server: concurrent connections

update to reflect changes in 3ebecd15615197b81efe79fcde6e4bedc0f6b09d

Signed-off-by: Gerhard Jörges <joerges@metratec.com>
